SOF-2298: Make repository transactions conform to ACID

KvelaGorrrrnio opened the pull request:
_**Note:** Since this is a major change to how the system operates, I encourage reviewers to have extra focus on the details in the PR - especially in the repository files._

This PR does 2 major things.

One is to make the Rundown, Segment, Part and Piece repositories to an aggregate structure, where everything goes through the Rundown repository. This is done to increase consistency between in-memory and database state, e.g. by allowing the cache decorator of the rundown aggregate repository to update the cache, even if a method call was on e.g, segment level. It also enables the Rundown aggregate repository to do optimizations like bulk writes for larger operations - which is also done in this PR.

The second major change is to make compound commands to the database for the Rundown hierarchy transactional. So if a portion of the command fails, then the entire thing is rolled back and reported as an error.
In conjunction with this, bulk writes is used to optimize the execution speed. In doing so, most methods in the segment, part, and piece repositories build mongo queries, instead of executing them. They can then be bundled and executed with the method `executeQueries`.

Additional changes made in the PR are:
- The cache decorators for segment and part repositories are removed, due to them being unused and the new rundown aggregate repository structure.
- The model type for a repository is moved from the `getCollection` to the class declaration, to fixate it for the entire class. This ensures that whenever `getCollection` is used, the correct model type is used instead of the reduced default document type `MongId`.
- Removes the `InMemoryMongo` database integration tests from the unit test suite. If they are needed, then we should implement them as integration-tests, e.g. with [Testcontainers](https://testcontainers.com/).
- Saving a rundown also removes orphaned content from that rundown in the database.
- Unplanned pieces are given unique ids.
- Rundown id is set on unplanned parts.

KvelaGorrrrnio committed:
chore: Deletes the unused class CachedStudioRepository.

KvelaGorrrrnio committed:
chore: Removes CachedSegmentRepository and CachedPartRepository.

KvelaGorrrrnio committed:
refactor: Move model type from BaseMongoRepository.getCollection method to the class.

KvelaGorrrrnio committed:
fix: Save flow for rundown, segment, part and piece uses transactions. Also removes the mongodb in memory tests.

RasmusAlbrektsen reviewed (commented):
Gotta say, i like it when you do ACID.

KvelaGorrrrnio committed:
fix: Delete-flow for rundown, segment, part and piece uses transactions.

KvelaGorrrrnio committed:
refactor: Changes RundownRepository to RundownAggregateRepository.

By having the rundown repository be an aggragate root, it is possible to control how the rundown hierarchy is persisted and the data consistency can be maintained by the aggregate root.

KvelaGorrrrnio committed:
refactor: Cleaned up in rundown hierarchy repository interfaces and methods, favoring query building.

KvelaGorrrrnio committed:
refactor: Moves transaction-based commands in rundown hierarchy repositories to the rundown aggregate repository.

KvelaGorrrrnio committed:
refactor: Use AnyBulkWriteOperation<Model> as return type for buildXQuery methods in repositories.

KvelaGorrrrnio committed:
refactor: MongoRundownAggregateRepository.getBasicRundowns uses map on the mongo cursor before converting it to an array.

KvelaGorrrrnio committed:
fix: Deletes pieces for unplanned parts when trying to delete unplanned parts for segment.

KvelaGorrrrnio committed:
refactor: Renames methods that removes all of an entity type to contain All in the name.

KvelaGorrrrnio committed:
Merge branch 'staging' into SOF-2298/make-repository-transactions-conform-to-acid

KvelaGorrrrnio committed:
fix: Ensures that unplanned pieces have unique ids.

KvelaGorrrrnio committed:
fix: Rundown id is assinged to unplanned parts.

KvelaGorrrrnio committed:
fix: Orphaned content is deleted from database when rundown is saved.

KvelaGorrrrnio committed:
Merge remote-tracking branch 'origin/SOF-2298/make-repository-transactions-conform-to-acid' into SOF-2298/make-repository-transactions-conform-to-acid

KvelaGorrrrnio committed:
fix: Adds infinite pieces to pieces-to-keep when saving rundown. Removes redundant repository delete calls and methods.

KvelaGorrrrnio committed:
refactor: RundownTimelineService.deleteUnsyncedSegments uses forEach instead of map.

KvelaGorrrrnio committed:
refactor: Removes PieceRepository.getPiecesFromIds since it was dead code.

KvelaGorrrrnio committed:
refactor: Removes unnecessary await.

KvelaGorrrrnio committed:
refactor: Removes dead code. Makes X_COLLECTION_NAME private for rundown hierarchy.

MadsThy commented on src/data-access/repositories/interfaces/piece-repository.ts:
```diff
@@ -1,12 +1 @@
-  deleteAllUnsyncedPieces(): Promise<void>
-  deleteAllUnplannedPieces(): Promise<void>
-}
+export interface PieceRepository {}
```
You cleaned house? :P

KvelaGorrrrnio replied:
I did :) This interface is now removed :)

MadsThy commented on src/data-access/repositories/interfaces/rundown-aggregate-repository.ts:
```diff
@@ -0,0 +1,6 @@
+import { PartRepository } from './part-repository'
+import { PieceRepository } from './piece-repository'
+
+export interface RundownAggregateRepository extends RundownRepository, SegmentRepository, PartRepository, PieceRepository {}
```
Why are you making so many empty types? xD

KvelaGorrrrnio replied:
This is not an empty type, since it combines multiple interfaces. I have though changed it to a type declaration instead :)

MadsThy commented on src/data-access/repositories/mongo/mongo-piece-repository.ts:
```diff
@@ -19,48 +20,70 @@ export class MongoPieceRepository extends BaseMongoRepository implements PieceRe
-    ).toArray()
-    return this.mongoEntityConverter.convertToPieces(mongoPieces)
+    return this.getCollection()
+      .find<MongoPiece>({_id: { $in: pieceIds } })
```
```suggestion
      .find<MongoPiece>({ _id: { $in: pieceIds } })
```

Formatting :D

KvelaGorrrrnio replied:
Formatting applied ✏️

LindvedKrvang commented on src/business-logic/services/execute-action-service.ts:
```diff
@@ -150,17 +150,19 @@ export class ExecuteActionService implements ActionService {
   }
 
-  private createPartFromAction(partAction: PartAction): Part {
+  private createPartFromAction(partAction: PartAction, rundownId: string): Part {
```
FYI: `PartAction` does have a `rundownId` already. However, only "rundown-specific" Actions have it, so it might not be useable for this usecase (assuming that the usecase here is to provide the RundownId of the Rundown the Part is being inserted into).

KvelaGorrrrnio replied:
I know, but as you state, we want to have the rundown id on all parts that are inserted into a rundown :)

LindvedKrvang commented on src/data-access/facades/repository-facade.ts:
```diff
@@ -82,13 +81,18 @@ export class RepositoryFacade {
+    return this.createRundownAggregateRepository()
+  }
+
+  public static createRundownAggregateRepository(): RundownAggregateRepository {
```
Should be private

KvelaGorrrrnio replied:
It should and now is :)

LindvedKrvang commented on src/data-access/repositories/cache/cached-rundown-aggregate-repository.ts:
```diff
@@ -0,0 +1,55 @@
+    this.cachedRundowns.delete(rundownId)
+  }
+
+  public getSegment(segmentId: string): Promise<Segment> {
+    return this.rundownAggregateRepository.getSegment(segmentId)
+  }
+
+  public getPart(partId: string): Promise<Part> {
+    return this.rundownAggregateRepository.getPart(partId)
+  }
```
We are not using the cache when fetching Segments and Parts?

KvelaGorrrrnio replied:
We are not :) Currently, they are only used when we receive a delete X event from Sofie core via MongoDB to get the rundown to which they belong. We could search through all cached rundowns to try and find the segment or part before asking the database?

LindvedKrvang commented on src/data-access/repositories/interfaces/piece-repository.ts:
```diff
@@ -1,12 +1 @@
-  deleteAllUnsyncedPieces(): Promise<void>
-  deleteAllUnplannedPieces(): Promise<void>
-}
+export interface PieceRepository {}
```
If it's empty, why do we still have it?

KvelaGorrrrnio replied:
Good question! It is now removed :)

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-action-trigger-repository.ts:
```diff
@@ -4,10 +4,11 @@ import { ActionTrigger } from '../../../model/entities/action-trigger'
 const ACTION_TRIGGER_COLLECTION: string = 'actionTriggers'
 
-export class MongoActionTriggerRepository extends BaseMongoRepository implements ActionTriggerRepository {
+export class MongoActionTriggerRepository extends BaseMongoRepository<ActionTrigger & MongoId> implements ActionTriggerRepository {
```
One of the reasons why the generic was on `getCollection` and not on the class of `BaseMongoRepository` was exactly for this reason that we didn't need to add `MongoId` on things where we don't need/use the MongoId syntax.

KvelaGorrrrnio replied:
By having it on the `getCollection` method, defining the type for what is saved to the database becomes optional. E.g. unless I as developer explicitly set up the type argument for the `getCollection` call, there is nothing stopping me from saving a Rundown to the ActionTrigger collection.
By having it on the class, no extra steps are needed from the developer and the type system helps with only saving valid data to the database.

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-ad-lib-actions-repository.ts:
```diff
@@ -6,6 +6,7 @@ import { BaseMongoRepository } from './base-mongo-repository'
 const AD_LIB_ACTIONS_COLLECTION: string = 'adLibActions'
 
 interface MongoAdLibAction {
+  _id: string
```
I'm quite sad to see that this is added when we don't use it.

KvelaGorrrrnio replied:
That doesn't contribute towards the domains OKR 😢 Personally I don't see the issue with having the identifier on a model which has unique instances. It is the contract for working with MongoDB.

LindvedKrvang replied:
Don't really know what OKR has to do with anything here..?

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-ad-lib-piece-repository.ts:
```diff
@@ -8,6 +8,7 @@ import { MongoId } from './mongo-entity-converter'
 const AD_LIB_PIECES_COLLECTION: string = 'adLibPieces'
 
 interface MongoAdLibPiece {
+  _id: string
```
Same comment as for `MongoAdLibAction`

KvelaGorrrrnio replied:
https://github.com/tv2/sofie-server/pull/249#discussion_r1802452972

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-entity-converter.ts:
```diff
@@ -134,6 +136,7 @@ export interface MongoTimeline extends MongoId {
 }
 
 export interface MongoStudio {
+  _id: string
```
Same comment about adding this id when it is not used. Goes for all the places where this happened in this PR

KvelaGorrrrnio replied:
https://github.com/tv2/sofie-server/pull/249#discussion_r1802452972

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-entity-converter.ts:
```diff
@@ -177,6 +181,18 @@ export interface MongoSystemInformation extends MongoId {
   name: string
 }
 
+export interface MongoAction extends MongoId {
```
A big reason for the `MongoX` interfaces is because Core saves entities in certains ways that we don't want our entities to have.

`Action` is not something that comes from Core so we don't need a `MongoAction` interface. We can simply use our `Action` interfacej.

Also, our Actions doesn't have the `_id` so it shouldn't extend `MongoId`.

KvelaGorrrrnio replied:
The biggest reason for having the MongoX interfaces is being able to map from a specific database format (currently MongoDB) to the data format used by the application. All documents we fetch from the database has the `_id` attribute, so for actions (and other models) the `_id` attribute is implicitly passed along. I have updated the MongoActionRepository to make use of the entity converter.

And the action repository uses the `_id` for grabbing actions.

LindvedKrvang replied:
As we talk about I don't see the need for it, but I guess it's what we are going with for now ;)

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-part-repository.ts:
```diff
@@ -1,21 +1,22 @@
+const PART_COLLECTION_NAME: string = 'executedParts' // TODO: Once we control ingest rename to "parts".
 
-export class MongoPartRepository extends BaseMongoRepository implements PartRepository {
+export class MongoPartRepository extends BaseMongoRepository<MongoPart> {
```
Why does this no longer extend `PartRepository`?
What's the point of having the abstraction if the implementation don't use it?

KvelaGorrrrnio replied:
It never has extended the interface 😉 But the reason being that it no longer implements the interface is that it doesn't conform to the interface. It is the RundownAggregateRepository that conforms to the repository interfaces.

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-rundown-aggregate-repository.ts:
```diff
@@ -0,0 +1,110 @@
+
+  constructor(
+    mongoDatabase: MongoDatabase,
+    private readonly mongoSegmentRepository: MongoSegmentRepository,
+    private readonly mongoPartRepository: MongoPartRepository,
+    private readonly mongoPieceRepository: MongoPieceRepository,
+    private readonly mongoEntityConverter: MongoEntityConverter,
```
We should depend on abstractions, not implementations.

KvelaGorrrrnio replied:
In most cases yes, but the Mongo repositories for segments, parts and pieces are under the rundown aggregate repository and should not be used outside of the aggregate. The functionality that they provide are exposed via the repository interfaces through the RundownAggregateRepository.

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-part-repository.ts:
```diff
@@ -1,21 +1,22 @@
+export class MongoPartRepository extends BaseMongoRepository<MongoPart> {
   constructor(
     mongoDatabase: MongoDatabase,
+    private readonly mongoPieceRepository: MongoPieceRepository,
```
You should depend on the abstraction, not the implementation.

KvelaGorrrrnio replied:
See https://github.com/tv2/sofie-server/pull/249#discussion_r1802459321.

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-part-repository.ts:
```diff
@@ -46,74 +47,103 @@ export class MongoPartRepository extends BaseMongoRepository implements PartRepo
-    this.assertDatabaseConnection(this.delete.name)
-    await this.pieceRepository.deletePiecesForPart(partId)
-    await this.getCollection().deleteMany({ _id: partId })
+  public getPartIdsForSegment(segmentId: string, mongoPart: Partial<MongoPart> = {}): Promise<readonly string[]> {
```
This method isn't being used.

KvelaGorrrrnio replied:
It is now gone.

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-part-repository.ts:
```diff
@@ -46,74 +47,103 @@ export class MongoPartRepository extends BaseMongoRepository implements PartRepo
-    this.assertDatabaseConnection(this.deleteUnsyncedPartsForSegment.name)
-    const unsyncedFilter: Partial<MongoPart> = { isUnsynced: true }
-    const unsyncedParts: Part[] = await this.getParts(segmentId, unsyncedFilter)
+  public buildDeletePartsForRundownQuery(rundownId: string): AnyBulkWriteOperation<MongoPart> {
```
This method isn't used

KvelaGorrrrnio replied:
It is now gone.

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-part-repository.ts:
```diff
@@ -46,74 +47,103 @@ export class MongoPartRepository extends BaseMongoRepository implements PartRepo
+  }
 
-    await Promise.all(unsyncedParts.map(async (part) => this.pieceRepository.deletePiecesForPart(part.id)))
+  public buildDeletePartsForSegmentQuery(segmentId: string, mongoPart: Partial<MongoPart> = {}): AnyBulkWriteOperation<MongoPart> {
```
This method isn't used

KvelaGorrrrnio replied:
![image](https://github.com/user-attachments/assets/b6f5814d-321e-4d44-b866-b1dbcb3f5539)

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-part-repository.ts:
```diff
@@ -46,74 +47,103 @@ export class MongoPartRepository extends BaseMongoRepository implements PartRepo
+  }
 
-    const partsDeletedResult: DeleteResult = await this.getCollection().deleteMany({ ...unsyncedFilter, segmentId: segmentId })
+  public buildDeletePartQuery(partId: string): AnyBulkWriteOperation<MongoPart> {
```
This method isn't used

KvelaGorrrrnio replied:
And it is time for it to go ⌛

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-part-repository.ts:
```diff
@@ -46,74 +47,103 @@ export class MongoPartRepository extends BaseMongoRepository implements PartRepo
 
-    if (!partsDeletedResult.acknowledged) {
-      throw new DeletionFailedException(`Deletion of parts was not acknowledged, for segmentId: ${segmentId}`)
+  public buildDeleteUnsyncedPartsForSegmentQuery(segmentId: string): AnyBulkWriteOperation<MongoPart> {
```
This method isn't used

KvelaGorrrrnio replied:
It is no longer with us 👼

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-part-repository.ts:
```diff
@@ -46,74 +47,103 @@ export class MongoPartRepository extends BaseMongoRepository implements PartRepo
+        filter: { isPlanned: false },
+      }
+    }
   }
```
These methods are not used

KvelaGorrrrnio replied:
And they are gone!

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-piece-repository.ts:
```diff
@@ -1,15 +1,16 @@
+const PIECE_COLLECTION_NAME: string = 'executedPieces' // TODO: Once we control ingest rename to "pieces".
 
-export class MongoPieceRepository extends BaseMongoRepository implements PieceRepository {
+export class MongoPieceRepository extends BaseMongoRepository<MongoPiece> {
```
What's the reason for this not implementing `PieceRepository`?

KvelaGorrrrnio replied:
See https://github.com/tv2/sofie-server/pull/249#discussion_r1802456012.

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-piece-repository.ts:
```diff
@@ -19,48 +20,70 @@ export class MongoPieceRepository extends BaseMongoRepository implements PieceRe
+    }
+  }
+
+  public buildDeletePiecesForPartQuery(partId: string): AnyBulkWriteOperation<MongoPiece> {
```
This method isn't used

KvelaGorrrrnio replied:
Is now removed.

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-piece-repository.ts:
```diff
@@ -19,48 +20,70 @@ export class MongoPieceRepository extends BaseMongoRepository implements PieceRe
   }
 
   public async deleteUnsyncedInfinitePiecesNotOnAnyRundown(): Promise<void> {
+    this.assertDatabaseConnection(this.deleteUnsyncedInfinitePiecesNotOnAnyRundown.name)
```
Can't comment directly on the method declaration since it didn't change :p
This method isn't used anymore

KvelaGorrrrnio replied:
Is now removed.

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-piece-repository.ts:
```diff
@@ -80,27 +103,29 @@ export class MongoPieceRepository extends BaseMongoRepository implements PieceRe
+        filter: { isPlanned: false }
+      }
+    }
   }
```
These methods aren't used anymore

KvelaGorrrrnio replied:
I really thought I had gone through the repositories and removed unused methods. But they are gone now :D

LindvedKrvang replied:
Are you using WebStorm for that? 
Atleast for me, WebStorm changes the color of unused methods so they stand out.

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-segment-repository.ts:
```diff
@@ -1,22 +1,22 @@
+const SEGMENT_COLLECTION_NAME: string = 'executedSegments' // TODO: Once we control ingest rename to "segments".
 
-export class MongoSegmentRepository extends BaseMongoRepository implements SegmentRepository {
+export class MongoSegmentRepository extends BaseMongoRepository<MongoSegment> {
```
Same question as for the other repositories: Why does it not implement `SegmentRepository` and why are you depending on the concrete implementation in the constructor?

KvelaGorrrrnio replied:
See https://github.com/tv2/sofie-server/pull/249#discussion_r1802456012.

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-segment-repository.ts:
```diff
@@ -34,75 +34,81 @@ export class MongoSegmentRepository extends BaseMongoRepository implements Segme
-    await Promise.all(segments.map(async (segment) => this.partRepository.deletePartsForSegment(segment.id)))
 
-    const segmentDeleteResult: DeleteResult = await this.getCollection().deleteMany({ ...unsyncedFilter, rundownId: rundownId })
+  public buildDeleteUnsyncedSegmentsForRundownQuery(rundownId: string): AnyBulkWriteOperation<MongoSegment> {
```
This method isn't used

KvelaGorrrrnio replied:
And it now no longer exists 💀

LindvedKrvang commented on src/data-access/repositories/mongo/mongo-segment-repository.ts:
```diff
@@ -34,75 +34,81 @@ export class MongoSegmentRepository extends BaseMongoRepository implements Segme
-    await this.getCollection().deleteMany({
-      isUnsynced: true
-    })
+  public buildDeleteAllUnsyncedSegmentsQuery(): AnyBulkWriteOperation<MongoSegment> {
```
This method isn't used

KvelaGorrrrnio replied:
And it now no longer exists 💀

LindvedKrvang commented on src/business-logic/services/rundown-timeline-service.ts:
```diff
@@ -180,8 +162,8 @@ export class RundownTimelineService implements RundownService {
 
-    await this.deleteUnsyncedPreviousPart(rundown)
-    await this.deleteUnsyncedSegments(rundown)
+    this.emitDeleteUnsyncedPreviousPart(rundown)
+    this.deleteUnsyncedSegments(rundown)
     await this.saveRundown(rundown)
```
A thing to note: 
Since `emitDeleteUnsyncedPreviousPart` is emitted before we try to save the Rundown; it's possible for us to to emit the delete event, fail the save of the Rundown, and then not actually delete the Part that we just emitted was deleted.


Another thing to note - the actual deletion unsynced Parts:
In these changes (the entire PR) we no longer make an "active effort" to delete unsynced Parts. They are still deleted, but only because the unsynced Part is removed from the Parts array of the Segment when we do `Rundown.takeNext()` (happens inside the private method `Rundown.takeNextCursor()`). Note, that we only ever try to remove unsynced Parts from a Segment when the Segment is in the _active_ cursor of the Rundown when starting the Take.

This may, or may not, lead to an egdecase where we have a Part that becomes unsynced, we do a Take away from it, the save Rundown flow _fails_ and we now have a rouge unsynced Part.

The above egdecase is just a thought scenario. I haven't checked if it's an actual issue. But I wanted to mention it :)

KvelaGorrrrnio replied:
Nice observations! I agree with them. It is the same approach we have for other operations on the rundown, where we emit events, build and persist the timeline before we store the rundown. So I think if this should be called after saving the rundown, so should all other event emitting. But that is balancing perceived performance vs data consistency :)

LindvedKrvang replied:
Agree. Leaving the comment unresolved for other people to also see it :)

LindvedKrvang reviewed (commented):
Good work. A much needed improvement :D 

Before I can approve, I need to hear your reasoning for some of the choices you made. I have made comments where applicable.

Note: Not a fan of moving the placement of the generic in `BaseMongoRepositoy`. It was chosen on the method so we weren't force to put `MongoId` on a lot of our entities that simply doesn't use it (as you have added a lot of places in this PR).
Not saying we should move it back, but I would prefer that we avoid adding an attribute that we don't use.

KvelaGorrrrnio committed:
chore: RepositoryFacade.createRundownAggregateRepository is made private.

KvelaGorrrrnio committed:
chore: Removes dead code from MongoSegmentRepository.

KvelaGorrrrnio committed:
chore: Removes dead code from MongoPieceRepository.

KvelaGorrrrnio committed:
chore: Removes dead code from MongoPartRepository.

KvelaGorrrrnio committed:
chore: Removes PieceRepository interface.

KvelaGorrrrnio committed:
refactor: RundownAggregateRepository interface is converted to a type declaration.

KvelaGorrrrnio committed:
lint: Added spacing after bracket start in MongoPieceRepository.

RasmusAlbrektsen reviewed (approved):
Except for the unused methods i think it looks very good! But Lindved seems to have caught every one i saw.
I must agree with Lindved that i'm confused about us not depending on abstractions for the mongo repositories, i'd say we shouldn't change the approach unless we decide it together, just to avoid having multiple approaches in the same repository.

KvelaGorrrrnio committed:
refactor: Updates MongoActionRepository to use bulk operation for saving actions.

getActions is renamed to getSystemActions.
MongoAction to Action conversion is added.

KvelaGorrrrnio committed:
refactor: ActionGenerationService.generateActionsForRundown takes a rundown instead rundownId.

This lowers the time spent on initial load and ensures that the showstyle variant used is updated.

LindvedKrvang commented on src/data-access/repositories/interfaces/rundown-aggregate-repository.ts:
```diff
@@ -1,6 +1,5 @@
-import { PieceRepository } from './piece-repository'
 
-export interface RundownAggregateRepository extends RundownRepository, SegmentRepository, PartRepository, PieceRepository {}
+export type RundownAggregateRepository = RundownRepository & SegmentRepository & PartRepository
```
Is it necessary to have `RundownAggregateRepository` be an actual type?
We couldn't just have?:
```
export class MongoRundownAggreaterepository implements RundownRepository, SegmentRepository, PartRepository {
  ...
}
```

KvelaGorrrrnio replied:
I introduced it to have one place to update it, e.g. if the piece repository interface needs to be added again. It is also used in the repository facade. If it needs to go, I suggest that it is done as boy-scouting :)
